### PR TITLE
refactor(Drawer): refactor drawer closeIcon

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/discussions/42828

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Drawer support hidden close button when closeIcon setting to null or false |
| 🇨🇳 Chinese | 抽屉组件支持通过设置 closeIcon 为 null 或 false 隐藏关闭按钮 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 4432621</samp>

This pull request adds a feature to the `Drawer` component that allows hiding the close button by setting the `closeIcon` prop to `null` or `false`. It also updates the documentation and the test cases accordingly.

### <samp>🤖 Generated by Copilot at 4432621</samp>

* Allow hiding close button by setting `closeIcon` to `null` or `false` ([link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-93e3b026986cf56379143412817a4077a256d276a567a550593b2526310d1a69L14-R14), [link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-93e3b026986cf56379143412817a4077a256d276a567a550593b2526310d1a69L31-R32), [link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-93e3b026986cf56379143412817a4077a256d276a567a550593b2526310d1a69L41-R53), [link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-7855c9de8eab66dde84ba7237f52b0fb88c917e3c8c78ada5fafa1e98c259721L49-R49), [link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-bb2fbbde54ac7a64e8f607e35d8482d3150594c6466b025b29949ba82b675152L48-R48))
* Memoize close icon value based on `closeIcon` and `closable` props using `React.useMemo` hook ([link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-93e3b026986cf56379143412817a4077a256d276a567a550593b2526310d1a69L41-R53))
* Add test case for new `closeIcon` and `closable` props behavior in `Drawer.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/42993/files?diff=unified&w=0#diff-2f59df76b42f52eb16232a5efb82e602e6a1b1cc126238a84971d00efa948df8R245-R288))

### CODE REVIEW

File: components/drawer/DrawerPanel.tsx [hunk 1]
```diff
@@ -38,9 +38,19 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
...
   } = props;
 
+  const mergedCloseIcon = React.useMemo(() => {
+    if (!closable) {
+      return null;
+    }
+    if (closeIcon === true) {
+      return <CloseOutlined />;
+    }
+    return closeIcon || <CloseOutlined />;
```
[the remark this answers is not in the document]
Comment (Wxh16144) [reply components/drawer/DrawerPanel.tsx:48]: 另外这里提一嘴, 是不是可以不用 React.useMemo ?

File: components/drawer/DrawerPanel.tsx [hunk 2]
```diff
@@ -28,8 +28,8 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
     title,
     footer,
     extra,
-    closable = true,
-    closeIcon = <CloseOutlined />,
+    closeIcon,
+    closable = closeIcon !== null && closeIcon !== false,
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:32]: 这里不用填，解构相互依赖是个坏味道。容易被 reorder 破坏

File: components/drawer/DrawerPanel.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kiner-tang) [reply components/drawer/DrawerPanel.tsx:48]: 我改一下，再加点 case

File: components/drawer/DrawerPanel.tsx (hunk 2 above)
Comment (kiner-tang) [reply components/drawer/DrawerPanel.tsx:32]: ok

File: components/drawer/DrawerPanel.tsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ export interface DrawerPanelProps {
   extra?: React.ReactNode;
 
   closable?: boolean;
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:13]: ts 定义标注一下推荐用 `closeIcon=false`

File: components/drawer/DrawerPanel.tsx [hunk 4]
```diff
@@ -38,21 +38,33 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
     children,
   } = props;
 
-  const closeIconNode = closable && (
+  const mergedCloseIcon = React.useMemo(() => {
+    if (!closable) {
+      return null;
+    }
+    if (closeIcon === true) {
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:45]: 这个和上面那个结构一样，重复定义 `<CloseOutlined />` 了

File: components/drawer/DrawerPanel.tsx [hunk 5]
```diff
@@ -38,21 +44,33 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
...
+    if (!closable) {
+      return null;
+    }
+    if (closeIcon === undefined || closeIcon === true) {
+      return <CloseOutlined />;
+    }
+    return closeIcon === null || closeIcon === false ? null : closeIcon;
+  }, [closeIcon, closable]);
+
+  const mergedClosable = mergedCloseIcon || ['string', 'number'].includes(typeof mergedCloseIcon);
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:57]: 这个逻辑有点拗口，上面 `mergedCloseIcon` 已经有 `closable` 判断了，下面还要再过一次。不如先算 closable，然后根据 closable 算 icon

File: components/drawer/DrawerPanel.tsx [hunk 6]
```diff
@@ -38,21 +44,39 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
     children,
   } = props;
 
-  const closeIconNode = closable && (
+  const mergedClosable = React.useMemo(() => {
+    if (typeof closable === 'boolean') {
+      return closable;
+    }
+
+    return ['string', 'number', 'undefined'].includes(typeof closeIcon) || !!closeIcon;
```
Comment (zombieJ) [inline components/drawer/DrawerPanel.tsx:52]: 太复杂了，不可关闭其实就这两种情况：

```tsx
return closeIcon !== null && closeIcon !== false
```

Comment (kiner-tang) [reply components/drawer/DrawerPanel.tsx]: 确实，我把问题复杂化了

File: components/drawer/index.zh-CN.md [hunk 7]
```diff
@@ -45,8 +45,7 @@ demo:
 | afterOpenChange | 切换抽屉时动画结束后的回调 | function(open) | - |  |
 | bodyStyle | 可用于设置 Drawer 内容部分的样式 | CSSProperties | - |  |
 | className | Drawer 容器外层 className 设置，如果需要设置最外层，请使用 rootClassName | string | - |  |
-| closable | 是否显示左上角的关闭按钮 | boolean | true |  |
-| closeIcon | 自定义关闭图标 | ReactNode | &lt;CloseOutlined /> |  |
+| closeIcon | 自定义关闭图标，5.7.0：设置为 null 或 false 时隐藏关闭按钮 | boolean \| ReactNode | &lt;CloseOutlined /> |  |
```
Comment (zombieJ) [inline components/drawer/index.zh-CN.md:48]: > 自定义关闭图标。`5.7.0` 后设置为 `null` 或 `false` ...

断一下。另外反引号忘了~

**Review by zombieJ — APPROVED** [review]

**Pull request merged**